Initial working example mocked lsf queue running poly

oysteoh opened the pull request:
In order for us to enable testing on different feedback from LSF we have decided to implement a mocked version enabling us to tweak on the responses and then verify how ERT interacts / handles different responses / errors. 

Based on this we verified the current behavior did not work as expected - and we ended up in an infitie loop inside job_monitor. By using the returned status from submit job and based on set status to `JobStatusType.JOB_QUEUE_DONE` . Then we will utilize the retry mechanism again in the queue itself when fail to submit ( which is currently a 100 x 2 sec schema ). 

The sad thing about this functionality which has been bypassed for a while and caused ERT just hanging when loosing connection to LSF is that the queue after the number of retries specified are throwing the util abort and ERT blows up. We should mitigate this and let jobs fail individually if LSF is unavailable. If i.e 5 out of 100 jobs fails due to lsf have some kind of hiccup that should be ok. 

Will spawn a new issue around this.

markusdregi commented:
This is going to do wonders for our possibility to work with the lsf-drivers 😎

markusdregi commented on res/job_queue/job_queue_node.py:
```diff
@@ -154,7 +154,10 @@ def _should_be_killed(self):
 
-        self.submit(driver)
+        submit_status = self.submit(driver)
+        if submit_status != 0:
```
This is a magic `0`. It would be preferable if we can expose the status enum in Python and use it instead...

markusdregi commented on tests/lsf/mocked_jobs/mock_bjobs:
```diff
@@ -0,0 +1,44 @@
+                )
+
+
+def _print_no_unfinished_jobs():
```
This does not seem to be used?

markusdregi commented on tests/lsf/mocked_jobs/mock_bjobs:
```diff
@@ -0,0 +1,44 @@
+
+timestamp = str(datetime.datetime.now())
+
+with open("job_paths", "r") as job_paths:
```
It would be neat if we could get the indentation level of this code down a bit. A good start would be to parse `job_paths` and close it again immediately :)

markusdregi commented on tests/lsf/test_lsf_driver.py:
```diff
@@ -0,0 +1,140 @@
+    shutil.copytree(
+        "mocked_jobs",
+        os.path.join(str(tmpdir), "poly_example"),
+        dirs_exist_ok=True,
```
Should we not fail if the directory already exist?

markusdregi commented on tests/lsf/test_lsf_driver.py:
```diff
@@ -0,0 +1,140 @@
+from ert_shared.main import ert_parser
+
+
+def test_run_mocked_lsf_queue(tmpdir, source_root):
```
Seems like we are doing quite some path operations here. Perhaps we could utilise `pathlib` for increased readability?

markusdregi commented on tests/lsf/test_lsf_driver.py:
```diff
@@ -0,0 +1,140 @@
+        dirs_exist_ok=True,
+    )
+
+    with tmpdir.as_cwd():
```
If we turn this into a pytest fixture it can be reused for both tests...

markusdregi commented on tests/lsf/test_lsf_driver.py:
```diff
@@ -0,0 +1,140 @@
+    """
+
+    script = dedent(
+        """#!/usr/bin/env python3
```
To not having to maintain two submission implementations I suggest we make the fail script only take responsibility for failing and then afterwards calling the `mock_bsub` script above.

markusdregi commented on res/job_queue/job_queue_node.py:
```diff
@@ -154,7 +154,10 @@ def _should_be_killed(self):
-        self.submit(driver)
+        submit_status = self.submit(driver)
+        if submit_status != 0:
+            self._set_status(JobStatusType.JOB_QUEUE_DONE)
```
This makes a lot of sense given the transition schema for the job queue documented in the C code.

markusdregi reviewed (commented):
I think the approach of the PR is solid! The tests are worth their weight in gold and the approach is sound 👏🏻 

I had some comments to clean up the code a bit...

markusdregi commented on tests/ert_tests/lsf_queue/test_lsf_driver.py:
```diff
@@ -0,0 +1,128 @@
+
+    run_cli(parsed)
+
+    assert True is True
```
If not we're really out of luck! 😁

oysteoh replied:
ah, yes, forgot that one... not sure how well we should assert at this point. Either ERT has exploded or succeeded when reaching this point. Should i traverse the runpaths and verify OK files or something? Or is it sufficient just by running without error?

markusdregi commented on tests/ert_tests/lsf_queue/test_lsf_driver.py:
```diff
@@ -0,0 +1,128 @@
+import subprocess
+
+num = random.random()
+if num < {percentage_successful_submits / 100}:
```
I also believe that the sign of comparison is the wrong way here. It has no effect since we only use it with `50`, but try injecting `100` here. Which we would expect make it always successful ;)

markusdregi reviewed (commented):
This is looking really neat 👏🏻

I left a few final comments...

I also made a local attempt at moving the entire construction into the fixture, including modifying the configuration and injecting the bsub wrapper script. But indirect pytest parameterisation is a rabbit hole I always enter with good intentions and slowly and scarred back out of 🤯

oysteoh commented:
> This is looking really neat 👏🏻
> 
> I left a few final comments...
> 
> I also made a local attempt at moving the entire construction into the fixture, including modifying the configuration and injecting the bsub wrapper script. But indirect pytest parameterisation is a rabbit hole I always enter with good intentions and slowly and scarred back out of 🤯

Probably ended up in the same rabbithole i was fighting my way out of again today 😂

oysteoh commented:
Squashed commits - fixed the comments you had - hopefully for the better. Also added a better description in the mocked jobs of what is going on so it is easier to pick this up later.

With regards to the test itself and the fixture you also tried doing - not sure if it is possible / make the code easier. As said - i also tried and i found it not to improve much. But i guess there are better ways than i managed to pan out! That said - hopefully the code is separated into more or less easy to understand / individual functions and it should be quite easy making new iterations on it making more tests !

markusdregi reviewed (approved):
This will do good 💯
